trie-db/src/lookup.rs

```diff
@@ -655,4 +668,175 @@ where
 		}
 		Ok(None)
 	}
+
+	/// Look up the merkle value (hash) of the node that is the closest descendant for the provided
+	/// key.
+	///
+	/// When the provided key leads to a node, then the merkle value of that node
+	/// is returned. However, if the key does not lead to a node, then the merkle value
+	/// of the closest descendant is returned. `None` if no such descendant exists.
+	fn lookup_first_descendent_without_cache(
+		mut self,
+		nibble_key: NibbleSlice,
+		full_key: &[u8],
+	) -> Result<Option<TrieHash<L>>, TrieHash<L>, CError<L>> {
+		let mut partial = nibble_key;
+		let mut hash = self.hash;
+		let mut key_nibbles = 0;
+
+		// this loop iterates through non-inline nodes.
+		for depth in 0.. {
+			let node_data = match self.db.get(&hash, nibble_key.mid(key_nibbles).left()) {
+				Some(value) => value,
+				None =>
+					return Err(Box::new(match depth {
+						0 => TrieError::InvalidStateRoot(hash),
+						_ => TrieError::IncompleteDatabase(hash),
+					})),
+			};
+
+			self.record(|| TrieAccess::EncodedNode {
+				hash,
+				encoded_node: node_data.as_slice().into(),
+			});
+
+			// this loop iterates through all inline children (usually max 1)
+			// without incrementing the depth.
+			let mut node_data = &node_data[..];
+			loop {
+				let decoded = match L::Codec::decode(node_data) {
+					Ok(node) => node,
+					Err(e) => return Err(Box::new(TrieError::DecoderError(hash, e))),
+				};
+
+				let next_node = match decoded {
+					Node::Leaf(slice, _) => {
+						// The leaf slice can be longer than remainder of the provided key
+						// (descendent), but not the other way around.
+						if !slice.starts_with(&partial) {
+							self.record(|| TrieAccess::NonExisting { full_key });
+							return Ok(None)
+						}
+
+						if partial.len() != slice.len() {
+							self.record(|| TrieAccess::NonExisting { full_key });
+						}
+
+						return Ok(Some(hash))
+					},
+					Node::Extension(slice, item) => {
+						let common_prefix_len = partial.common_prefix(&slice);
+
+						let start_matches = if partial.len() <= slice.len() {
+							// Extension slice can be longer than remainder of the provided key
+							// (descendent), ensure the extension slice starts with the remainder
+							// of the provided key.
+							//
+							// This effectively returns the hash of the `Node::Branch` at the next
+							// iteration.
+							common_prefix_len == partial.len()
```

Comment:
Then in case the key is in the middle of the extension partial and match all along, the return merkle value should be the one of the extension.
But tbh, I would just put all `extension` related as unimplemented (and remove the variant from the tests).

Reply:
That makes sense! Have changed the code to also use `starts_with`.
Have kept the `extension` for not exposing a variation of `test_layouts!` but could place this under an `unimplemented` as well, let me know which one you think is better here 🙏